File: src/main/java/com/cloudogu/gitops/utils/DockerImageParser.java
package com.cloudogu.gitops.utils;

import lombok.RequiredArgsConstructor;

public class DockerImageParser {

private static final int MIN_SEGMENTS_WITH_REGISTRY = 2;
private static final int REPOSITORY_SEGMENT_COUNT = 2;

@lombok.Getter
@RequiredArgsConstructor
public static class Image {
	private final String registry;
	private final String repository;
	private final String tag;

	public String getRegistryAndRepositoryAsString() {
	if (registry == null || registry.isEmpty()) {
		return repository;
	}
	return registry + "/" + repository;
	}

	@Override
	public String toString() {
	return getRegistryAndRepositoryAsString() + ":" + tag;
	}
}

public static Image parse(String image) {
	int lastSlash = image.lastIndexOf('/');
	int lastColon = image.lastIndexOf(':');
	if (lastColon == -1 || lastColon < lastSlash) {
	throw new IllegalArgumentException(
		"Cannot set image '"
			+ image
			+ "' due to missing tag. Must be the format '$repository:$tag'");
	}

	ImageAndTag tuple = splitTag(image);
	String imageWithoutTag = tuple.imageWithoutTag();
	String tag = tuple.tag();

	String[] parts = imageWithoutTag.split("/");
	String repository;
	String registry;

	if (parts.length >= MIN_SEGMENTS_WITH_REGISTRY) {
	repository = parts[parts.length - REPOSITORY_SEGMENT_COUNT] + "/" + parts[parts.length - 1];
	StringBuilder registryBuilder = new StringBuilder();
	for (int i = 0; i < parts.length - REPOSITORY_SEGMENT_COUNT; i++) {
		if (i > 0) {
		registryBuilder.append("/");
		}
		registryBuilder.append(parts[i]);
	}
	registry = registryBuilder.toString();
	} else {
	repository = imageWithoutTag;
	registry = "";
	}

	return new Image(registry, repository, tag);
}

private static ImageAndTag splitTag(String image) {
	int lastColon = image.lastIndexOf(':');
	String imageWithoutTag = image.substring(0, lastColon);
	String tag = image.substring(lastColon + 1);
	return new ImageAndTag(imageWithoutTag, tag);
}

private record ImageAndTag(String imageWithoutTag, String tag) {}
}

File: src/main/kotlin/com/cloudogu/gitops/utils/DockerImageParser.kt
package com.cloudogu.gitops.utils

class DockerImageParser {
    class Image(
        val registry: String,
        val repository: String,
        val tag: String
    ) {
        fun getRegistryAndRepositoryAsString(): String {
            if (registry.isEmpty()) {
                return repository
            }
            return "$registry/$repository"
        }

        override fun toString(): String {
            return "${getRegistryAndRepositoryAsString()}:$tag"
        }
    }

    companion object {
        @JvmStatic
        fun parse(image: String): Image {
            if (!image.contains(":")) {
                // Most helm charts expect an explicit image tag, otherwise they use the version set by the app.
                // This will likely be unexpected so force using a tag
                throw RuntimeException("Cannot set image '$image' due to missing tag. Must be the format '\$repository:\$tag'")
            }

            val tuple = splitTag(image)
            val imageWithoutTag = tuple.first
            val tag = tuple.second

            val parts = imageWithoutTag.split("/").toMutableList()
            val repository = parts.takeLast(2).joinToString("/")
            
            // Drop last 2 elements to get the registry part
            val registry = if (parts.size > 2) {
                parts.dropLast(2).joinToString("/")
            } else {
                ""
            }

            return Image(registry, repository, tag)
        }

        private fun splitTag(image: String): Pair<String, String> {
            val imageParts = image.split(":")
            val tag = imageParts.last()
            val imageWithoutTag = imageParts.dropLast(1).joinToString(":")
            return Pair(imageWithoutTag, tag)
        }
    }
}
